apps/system/js/audio_channel_controller.js

```diff
@@ -0,0 +1,241 @@
+/* global BaseUI */
+'use strict';
+
+(function(exports) {
+  var FADE_IN_VOLUME = 1;
+  var FADE_OUT_VOLUME = 0.2;
+  var VIBRATION_DURATION = 1000;
+
+  /**
+   * AudioChannelController controls the audio channel to
+   * play or pause.
+   */
+  var AudioChannelController = function(app, channel) {
+    this.app = app;
+    this.name = channel.name;
+    this._channel = channel;
+    this._states = {
+      active: false,
+      playing: false,
+      fadingOut: false,
+      vibrating: false
+    };
+    this._policy = {};
+    this._generateID();
+    channel.addEventListener('activestatechanged', this);
+    app.element.addEventListener('_willdestroy', this);
+  };
+
+  AudioChannelController.prototype = Object.create(BaseUI.prototype); 
+
+  AudioChannelController.prototype.EVENT_PREFIX = 'audiochannel';
+
+  /**
+   * General event handler interface.
+   *
+   * @param {Event} evt The event to handle.
+   */
+  AudioChannelController.prototype.handleEvent = function(evt) {
+    switch (evt.type) {
+      case 'activestatechanged':
+        var channel = evt.target;
+        // TODO: We should get the `isActive` state from evt.isActive.
+        // Then we don't need to do `channel.isActive()` here.
+        channel.isActive().onsuccess = (evt) => {
+          this._states.active = evt.target.result;
+          this.publish('statechanged');
+        };
+        break;
+
+      // There cannot be a `_terminated` event for a window element.
+      // Because the element doesn't exist
+      // when the `_terminated` event is triggered.
+      case '_willdestroy':
+        this.publish('terminated');
+        break;
+    }
+  };  
+
+  /**
+   * Set the policy for handling this audio channel.
+   *
+   * @param {Object} policy Policy to handle this audio channel.
+   * @param {Boolean} [policy.isAllowedToPlay] Play or pause it.
+   * @param {Boolean} [policy.isNeededToFadeOut] Fade out or not.
+   * @param {Boolean} [policy.isNeededToVibrate] Vibrate or not.
+   * @param {Boolean} [policy.isNeededToResumeWhenOtherEnds]
+   * Resume when other audio channel ends if as true.
+   * @return {AudioChannelController}
+   */
+  AudioChannelController.prototype.setPolicy = function(policy) {
+    this._policy = policy || this._policy;
+    return this;
+  };
+
+  /**
+   * Handle the audio channel with the policy,
+   * and clear the policy.
+   *
+   * @param {Function} callback Do something
+   * after the audio channel is handled by the policy.
+   */
+  AudioChannelController.prototype.proceedPolicy = function(callback) {
+    var policy = this._policy;
+    if (policy.isAllowedToPlay) {
+      this._play();
+    } else if (policy.isAllowedToPlay != null) {
+      this._pause();
+    }
+    if (policy.isNeededToFadeOut) {
+      this._fadeOut();
+    } else if (policy.isNeededToFadeOut != null) {
+      this._fadeIn();
+    }
+    if (policy.isNeededToVibrate) {
+      this._vibrate();
+    }
+    callback && callback(policy);
```

Comment:
nit: if we really want to live with callback, return a promise (even it is resolved right away)